fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMutuallyByRecords.java

```diff
@@ -0,0 +1,529 @@
+                    .thenApply(vignore -> hasMore.get() ?
+                                          lastResult.get().get().getPrimaryKey() :
+                                          rangeEnd)
+                    .thenCompose(cont -> insertRanges(store.ensureContextActive(), targetRangeSets, packOrNull(rangeStart), packOrNull(cont))
+                            .thenApply(ignore -> !allRangesExhausted(cont, rangeEnd)));
+        });
+    }
+
+    @Nullable
+    private static Range fullyUnBuiltRange(List<Range> missingRanges, Range fragmentRange) {
+        // Return the fragmentRange 'as is' if it appears to be fully un-built in the
+        // missing ranges list. Else return null.
+        // * Assuming that the missing ranges are sorted and squashed
+        // * Optimize for a heavily fragmented missing ranges. (Should we perform a binary search?)
+        for (Range range: missingRanges) {
+            if (ByteArrayUtil.compareUnsigned(range.end, fragmentRange.end) >= 0) {
+                // missing range is beyond fragment
+                if (ByteArrayUtil.compareUnsigned(range.begin, fragmentRange.begin) <= 0) {
+                    return notEmptyRange(fragmentRange);
+                }
+                break;
+            }
+        }
+        return null;
+    }
+
+    @Nullable
+    private static Range partlyUnBuiltRange(List<Range> missingRanges, Range fragmentRange) {
+        // Return null if the range appears to be fully built in the missing ranges list. Else
+        // return the un-built range of (fragmentRange ∩ missingRanges)
+        // * Assuming that the missing ranges are sorted and squashed
+        // * Less optimized than the fullyUnBuiltRanges (this function is expected to be called only a handful
+        //   of times at the end of the indexing process)
+        for (Range range: missingRanges) {
+            if (ByteArrayUtil.compareUnsigned(range.begin, fragmentRange.end) > 0 ||
+                    ByteArrayUtil.compareUnsigned(range.end, fragmentRange.begin) < 0) {
+                // no range overlap
+                continue;
+            }
+            // Here: some range overlap, shrink to fragmentRange ∩ range
+            byte[] begin = ByteArrayUtil.compareUnsigned(range.begin, fragmentRange.begin) >= 0 ?
+                           range.begin : fragmentRange.begin;
+            byte[] end = ByteArrayUtil.compareUnsigned(range.end, fragmentRange.end) <= 0 ?
+                         range.end : fragmentRange.end;
+            if (ByteArrayUtil.compareUnsigned(begin, end) < 0) {
+                // if non-empty range
+                return notEmptyRange(new Range(begin, end));
+            }
+        }
+        return null;
+    }
+
+    @Nullable
+    private static Range notEmptyRange(Range range) {
+        // check the (rare, maybe test only) case of empty ranges
+        return ByteArrayUtil.compareUnsigned(range.begin, range.end) >= 0 ?
```

Comment:
The endpoints of a range are inclusive. If `begin==end` wouldn't that mean indexing exactly one key?

Reply:
Shouldn't the ranges overlap? (I.E. ending up as `[(a-b), (b-c), (c-d)...]`. )
In this case, in `[(a-b), (b-c), (c-c), (c-d)...]` the `(c-c)` range can probably be ignored.

Reply:
It is my understanding that there always can be something between `b` and `b'`, hence it must be included in the built-ranges set. Is that correct?

Reply:
If anyone adds a key of 3 bytes array - `[1, null, 7]` -  where will it go?

Reply:
Like @ScottDugas said, the range `[\x01, \x01\x00)` (inclusive at the beginning, exclusive at the end) contains only one key, namely `\x01`. If you want proof that for any byte string `b`, the range `[b, b+\x00)` contains only `b`, consider a different byte string `c` where `c != b`. If there exists an index `i` such that `c[i] != b[i]`, then pick the first such index. If `c[i] < b[i]`, then `c < b` so it is not in the range; otherwise, if `c[i] > b[i]`, then `c > b+\x00`. In the case where there is no such `i`, then it follows that either `c` is a prefix of `b`, in which case `c < b` and is not in the range, or `b` is a prefix of `c`, so `len(b)` is a valid index in `c`, and `c[len(b)] >= \x00` because it is an unsigned byte, so `c >= b+\x00` and `c` is therefore not in the range.

I'm not quite sure what you mean by "[1, null, 7]" as byte arrays can't have nulls in the middle. In any case, byte arrays are compared using unsigned lexicographic comparison, so the byte string `\x01\x00\x07` would sort after `\x01\x00`, and thus would not be in `[\x01, \x01\x00)`.

You _can_ have nulls in the middle of tuples. They get encoded as zero byte. And, in particular, the tuple `(1, null, 7)` would be encoded as `\x15\x01\x00\x15\x07`. The tuple `(1,)` is encoded as `\x15\x01`, and if you wanted to represent the range containing only the tuple `(1,)` would be `[\x15\x01, \x15\x01\x00)`, which again does not contain `(1, null, 7)` because that key is greater than the endpoint. Note that the tuple `(1, null)` encodes as `\x15\x01\x00`, which _is_ the endpoint, but because the endpoint is exclusive, it is also not in the range.

In any case, the endpoints on range sets are inclusive/exclusive, so that adding range `[a, b)` and `[b, c)` coalesces to `[a, c)` without any gaps. This is to map the way FDB ranges work. I guess `[a, a)` could be construed to be the empty range, though that seems like an error in the notation.